src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,431 @@
+	// ENCDataVault 1024 bits vault
+	{ "$encdv$3$4$bc93a92cf625e360$fe120bfb$739f9d85964cef2f63b927ff77f3328cc5192014ae21c29954c322f4e808fe5c8abe64cc150dbcb08cb334f3b5c28357f10d8d5c6a103e2c899402136c14633aaf8a5347959b33b80ade2e3f5698864605940dc1423704999e5da859d6584491bbe940f00f162c75d2766b868ef2b4c6bf599a2e8ccea3f7cdfab193744a8d10", "123456789ABCDEf" },
+	{ "$encdv$3$4$e9786d82101f30ef$d690010a$5748af61f246c1569cc72cbd6c97e0be46b1612842fd3c6284637af5351508a602c731739acb056b845dfa2b5befd40f14136b4336e0c8e98144555a90befbc85abfc4069fbb71709fa39e29f8c6a98d28e14251e50ffdd43d75252de2b0b14c386e72927c62ae39ba18ff8b32a339d882fdd0b15284af246fa50b7c6992f783", "Verylongbutstillfailpassword" },
+	{ NULL }
+};
+
+static int *cracked;
+static int any_cracked;
+static size_t cracked_size;
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+
+static struct custom_salt {
+	unsigned int version;
+	unsigned int algo_id;
+	unsigned char salt[ENC_SALT_SIZE];
+	unsigned char iv[ENC_IV_SIZE];
+	unsigned char encrypted_data[ENC_BLOCK_SIZE];
+	unsigned char keychain[ENC_KEYCHAIN_SIZE];
+} *cur_salt;
+
+void static enc_aes_ctr_iterated(const unsigned char *in, unsigned char *out, const unsigned char *key,
+                                 buffer_128 ivs[ENC_MAX_KEY_NUM], size_t len, size_t nb_keys, uint64_t counter)
+{
+	AES_KEY aes_key;
+	buffer_128 tmp_iv;
+	buffer_128 tmp_out;
+	int i, j;
+
+	AES_set_encrypt_key(key, ENC_KEY_SIZE * 8, &aes_key);
+	len >>= 4;
+	counter <<= 56;
+
+	for (i = 0; i < len; i++) {
+		tmp_iv.u64[0] = ivs[0].u64[0];
+		tmp_iv.u64[1] = counter;
+		AES_encrypt(tmp_iv.u8, tmp_iv.u8, &aes_key);
+		memcpy(tmp_out.u8, in, AES_BLOCK_SIZE);
+		enc_xor_block(tmp_out.u64, tmp_iv.u64);
+
+		for (j = 1; j < nb_keys; j++) {
+			tmp_iv.u64[0] = ivs[j].u64[0];
+			tmp_iv.u64[1] = counter;
+			AES_encrypt(tmp_iv.u8, tmp_iv.u8, &aes_key);
+			enc_xor_block(tmp_out.u64, tmp_iv.u64);
+		}
+		memcpy(out, tmp_out.u8, AES_BLOCK_SIZE);
+
+		// Increment counter, only 255 block encryption is supported.
+		counter += 0x0100000000000000;
+		out += AES_BLOCK_SIZE;
+		in += AES_BLOCK_SIZE;
+	}
+}
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	any_cracked = 0;
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto err;
+	if (!isdec(p))
+		goto err;
+	int version = atoi(p);
+	if (version < 1 || version > 3) {
```

Comment:
You allow version 2 now? Why?

Reply:
Sorry I was thinking of algo id again. I corrected it.